go/tasks/pluginmachinery/google/gke_token_source.go

```diff
@@ -0,0 +1,212 @@
+		})
+
+		m.deletion.Store(false)
+	}
+
+	return nil, false
+}
+
+func hasExpired(token *oauth2.Token) bool {
+	return token.Expiry.Round(0).Add(-defaultGracePeriod).Before(time.Now())
+}
+
+func (m *gkeTokenSource) getK8sServiceAccountToken(ctx context.Context, identity Identity) (string, error) {
+	serviceAccounts := m.kubeClient.CoreV1().ServiceAccounts(identity.K8sNamespace)
+	createTokenResponse, err := serviceAccounts.CreateToken(ctx, identity.K8sServiceAccount, &v1.TokenRequest{
+		Spec: v1.TokenRequestSpec{
+			Audiences: []string{
+				m.identityNamespace,
+			},
+		},
+	}, metav1.CreateOptions{})
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create k8s service account token")
+	}
+
+	return createTokenResponse.Status.Token, nil
+}
+
+func (m *gkeTokenSource) getGcpServiceAccount(ctx context.Context, identity Identity) (string, error) {
+	serviceAccounts := m.kubeClient.CoreV1().ServiceAccounts(identity.K8sNamespace)
+
+	serviceAccountResponse, err := serviceAccounts.Get(ctx, identity.K8sServiceAccount, metav1.GetOptions{})
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create k8s service account token")
+	}
+
+	for key, value := range serviceAccountResponse.Annotations {
+		if key == gcpServiceAccountAnnotationKey {
+			return value, nil
+		}
+	}
+
+	return "", errors.Errorf(
+		"[%v] annotation doesn't exist on k8s service account [%v/%v], read more at %v",
+		gcpServiceAccountAnnotationKey,
+		identity.K8sNamespace,
+		identity.K8sServiceAccount,
+		workflowIdentityDocURL)
+}
+
+func (m gkeTokenSource) GetTokenSource(ctx context.Context, identity Identity) (oauth2.TokenSource, error) {
+	if identity.K8sServiceAccount == "" {
+		identity.K8sServiceAccount = "default"
+	}
+
+	if cachedToken, ok := m.getCachedToken(ctx, identity); ok {
+		return oauth2.StaticTokenSource(cachedToken), nil
+	}
+
+	// when tokens expire, of we hit a miss, singleflight will do a most one request per SA
+	value, err := m.singleflight.Do(identity.K8sNamespace+"/"+identity.K8sServiceAccount, func() (interface{}, error) {
+		k8sServiceAccountToken, err := m.getK8sServiceAccountToken(ctx, identity)
+
+		if err != nil {
+			return nil, err
+		}
+
+		gcpServiceAccount, err := m.getGcpServiceAccount(ctx, identity)
+
+		if err != nil {
+			return nil, err
+		}
+
+		token, err := ExchangeToken(ctx, StsRequest{
+			SubjectToken:           k8sServiceAccountToken,
+			Scope:                  m.scope,
+			ServiceAccount:         gcpServiceAccount,
+			IdentityNamespace:      m.identityNamespace,
+			iamCredentialsEndpoint: m.iamCredentialsEndpoint,
+			federatedTokenEndpoint: m.federatedTokenEndpoint,
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		m.tokens.Store(identity, token)
+
+		return token, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return oauth2.StaticTokenSource(value.(*oauth2.Token)), nil
+}
+
+func getKubeClient(kubeConfigPath string, kubeConfig KubeClientConfig) (*kubernetes.Clientset, error) {
```

Comment:
you are using the webAPI, why do you need a kube client?

Reply:
i understand now, for the token stuff. We should pass this in in the task context

Reply:
@kumare3 have you seen my last comment? this code uses the REST client, and not the client we have in the task context. Not sure what to do about it